tensorpipe/common/tensor.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#if TENSORPIPE_HAS_CUDA_IPC_CHANNEL // TODO: Generic TENSORPIPE_HAS_CUDA.
+#include <cuda_runtime.h>
+#endif // TENSORPIPE_HAS_CUDA_IPC_CHANNEL
+
+namespace tensorpipe {
+
+enum class DeviceType {
+  kCpu = 0,
+#if TENSORPIPE_HAS_CUDA_IPC_CHANNEL
+  kCuda,
+#endif // TENSORPIPE_HAS_CUDA_IPC_CHANNEL
+};
+
+struct CpuTensor {
+  void* data{nullptr};
+  size_t length{0};
+};
+
+#if TENSORPIPE_HAS_CUDA_IPC_CHANNEL
+struct CudaTensor {
+  void* data{nullptr};
+  size_t length{0};
+  cudaStream_t stream{cudaStreamDefault};
+};
+#endif // TENSORPIPE_HAS_CUDA_IPC_CHANNEL
+
+struct Tensor {
```

Comment:
`std::variant` is C++17 and we've been limiting ourselves to C++14 so far. We could probably backport it, but we don't have that many unions laying around so it might not be worth.

Reply:
Right now we must support C++14 because PyTorch does but eventually I believe we'll drop it and move to C++17 and it would be nice to use the most idiomatic data structure once we do do. Having a polyfill for std::variant would allow us basically to do a drop-in replacement with limited user disruption. But since some disruption will still exist, I also see the argument for rolling our own type here.

One consideration is that I have never used std::variant and thus don't know how comfortable it is, both in terms of readability (for our code and our users' one) and in terms of static safety (getting compile errors when we forget to handle a case, and stuff like that).

Would it make sense that, before landing this diff, we start writing the PyTorch PR that migrates PyTorch to this new API, and see how it looks like? And, similarly, also stack on top of this another PR that makes the pipe able to handle CUDA channels and see if it's easy to do? Or, if this is too much work, find some other way to compare how std::variant and our own type would compare.

Reply:
Also, it is currently a bit "redundant" to have a member field called `tensor` inside the `Message::Tensor` class, which is of type `FooTensor`. I'm wondering if we could find a different naming. Maybe call the field `handle` and use `CpuHandle` and `CudaHandle`? Open to suggestions...